### DISCUSSION TITLE
httpasyncclient trace association error [Bug]

### DISCUSSION THREAD

**Discussion by hucaicai** [apache/skywalking#10828] [Q&A]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

When calling concurrently, there is a probability of triggering, and different trace associations are disordered.
<img width="857" alt="image" src="https://github.com/apache/skywalking/assets/46130847/2dfc60fe-8c01-4183-aa1d-398e1e82fa51">

### What you expected to happen

When HttpAsyncRequestExecutorInterceptor or SessionRequestCompleteInterceptor creates a createLocalSpan, the current context is for a request that has not yet ended and has not received a response result, so it is associated with the trace of the previous request; The underlying reason is that httpasyncclient uses the reactor model, and Thread will continuously fetch and process tasks. This problem occurs when processing other request tasks before the previous response result arrives.

### How to reproduce

The HTTP interface on the server has a Thread sleep of 1 second (increasing the probability of occurrence), and the client has a high probability of multiple threads calling concurrently

Using httpasyncclient version 4.1.5

```
// client

    ExecutorService executorService = Executors.newFixedThreadPool(10);
    for (int i = 0; i < 100000; i++) {
        executorService.submit(()->{
                httpclient.execute(post, new FutureCallback<HttpResponse>() {
                ...
        });
    }
```

        

```
 // server
    @RequestMapping("hello")
    public String hello() {
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return "hello";
    }
```

   

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
Async is always `disordered`. Maybe UI could consider ordering by start timestamp, but it is only a polish if the UI hasn't.
Anyway, this is not a bug.

**Reply by hucaicai** [reply]
that's right display
<img width="803" alt="image" src="https://github.com/apache/skywalking/assets/46130847/2f29b934-0078-402a-b7b0-0105cb03c9d4">

Two HttpAsyncClient/local reasons.
<img width="1433" alt="image" src="https://github.com/apache/skywalking/assets/46130847/4dad5b9b-dbd0-4deb-8fe8-3f2740a811df">

**Reply by hucaicai** [reply]
this is a bug. This is caused by the disordered association between two traces, and I am quite certain because I only called them once

**Reply by wu-sheng** [reply]
> Maybe UI could consider ordering by start timestamp, but it is only a polish if the UI hasn't.

The bug is a thing already done but doesn't work, this is a thing never added. 
This is common and why new PR would be accepted. If you want to add through a pull request, welcome.

**Reply by hucaicai** [reply]
This is not a problem with the UI, it is a problem with the data. Trace1 did not end, and the thread of trace1 executed trace2, causing trace2 to be associated with trace1.

**Reply by wu-sheng** [reply]
I can't agree. Different threads in Java don't share anything but  Reference to parent segment only

**Reply by hucaicai** [reply]
HttpAsyncRequestExecutorInterceptor only has createSpan and no stopSpan. StopSpan is called upon receiving a response.  
When no response is received, the thread will execute other traces, causing trace2 to be associated with trace1.

**Reply by hucaicai** [reply]
httpasyncclient  is Reactor model. CreateSpan completed, will proceed to perform other tasks

**Reply by wu-sheng** [reply]
I am unable to follow. Most plugins in SkyWalking are only contributed and tested through CI.
What you described is not related to backend even it is true.
I will leave this discussion here until someone knows the plugin join.
